modules/videoio/src/cap_ffmpeg_hw.hpp

```diff
@@ -0,0 +1,188 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2020-2021 Intel Corporation
+
+#include "opencv2/videoio.hpp"
+#include <list>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+#include <libavcodec/avcodec.h>
+#include <libavutil/avutil.h>
+#ifdef __cplusplus
+}
+#endif
+
+#define HW_FRAMES_POOL_SIZE     20
+
+using namespace cv;
+
+AVBufferRef* hw_create_device(VideoAccelerationType va_type, int hw_device);
+AVBufferRef* hw_create_frames(AVBufferRef *hw_device_ctx, int width, int height, AVPixelFormat hw_format, AVPixelFormat sw_format, int pool_size = HW_FRAMES_POOL_SIZE);
+AVCodec *hw_find_codec(AVCodecID id, AVBufferRef *hw_device_ctx, int (*check_category)(const AVCodec *), AVPixelFormat *hw_pix_fmt = NULL);
+AVPixelFormat hw_get_format_callback(struct AVCodecContext *ctx, const enum AVPixelFormat * fmt);
+
+#if LIBAVUTIL_VERSION_MAJOR >= 56 // FFMPEG 4.0+
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+#include <libavutil/hwcontext.h>
+#ifdef __cplusplus
+}
+#endif
+
+static AVHWDeviceType VideoAccelerationTypeToFFMPEG(VideoAccelerationType va_type) {
+    struct HWTypeFFMPEG {
+        VideoAccelerationType va_type;
+        AVHWDeviceType ffmpeg_type;
+    } ffmpeg_hw_types[] = {
+        { VIDEO_ACCELERATION_D3D11, AV_HWDEVICE_TYPE_D3D11VA },
+        { VIDEO_ACCELERATION_VAAPI, AV_HWDEVICE_TYPE_VAAPI },
+        { VIDEO_ACCELERATION_MFX, AV_HWDEVICE_TYPE_QSV }
+    };
+    for (const HWTypeFFMPEG& hw : ffmpeg_hw_types) {
+        if (va_type == hw.va_type)
+            return hw.ffmpeg_type;
+    }
+    return AV_HWDEVICE_TYPE_NONE;
+}
+
+AVBufferRef* hw_create_device(VideoAccelerationType va_type, int hw_device) {
+    AVHWDeviceType hw_type = VideoAccelerationTypeToFFMPEG(va_type);
+    if (AV_HWDEVICE_TYPE_NONE == hw_type)
+        return NULL;
+
+    AVBufferRef* hw_device_ctx = NULL;
+    char device[128] = "";
+    char* pdevice = NULL;
+    AVDictionary* options = NULL;
+    if (hw_device >= 0 && hw_device < 100000) {
+#ifdef _WIN32
+        snprintf(device, sizeof(device), "%d", hw_device);
+#else
+        snprintf(device, sizeof(device), "/dev/dri/renderD%d", 128 + hw_device);
+#endif
+        if (hw_type == AV_HWDEVICE_TYPE_QSV) {
+            av_dict_set(&options, "child_device", device, 0);
+        }
+        else {
+            pdevice = device;
+        }
+    }
+    av_hwdevice_ctx_create(&hw_device_ctx, hw_type, pdevice, options, 0);
```

Comment:
I observed weird behavior in this configuration:
- Windows laptop: i5-1035G1 (IceLake) + Intel iGPU + NVIDIA dGPU
- crash during `av_hwdevice_ctx_create()` call (`OPENCV_FFMPEG_DEBUG=1` for extra messages):
```
[OPENCV:FFMPEG:40] Using D3D9Ex device.
[OPENCV:FFMPEG:40] Initialize MFX session: API version is 1.34, implementation version is 1.31
[OPENCV:FFMPEG:16] Error setting child device handle: -6
... crash ...
```
- the last message from FFmpeg is from [here](https://github.com/FFmpeg/FFmpeg/blob/ba2cebb49ca974f44e5729975717f550a49aaa0f/libavutil/hwcontext_qsv.c#L1199)
- after debugging and checking that standalone `ffmpeg` tool works well with `qsv`, I have found a workaround with disabled OpenCL subsystem in OpenCV to bypass this crash
- next possible steps is to try to re-build FFmpeg with debug symbols and D3D/MFX support (current used binaries don't have them). FFmpeg is not very friendly with Windows build tools, so this may be tricky.
- anyway it would be hard to bypass MFX/OpenCL conflict in OpenCV
- System has the "latest" driver from MS/vendor 26.20.100.7870 (from 2/5/2020, ~1 year ago). Will try to install fresh driver from Intel resources.

Reply:
Thanks for checking on dual-GPU system.
It's small but annoying bug in ffmpeg (MFXClose called twice if non-Intel GPU), I will submit [FFMpeg patch](https://github.com/FFmpeg/FFmpeg/compare/master...mikhail-nikolskiy:fix-qsv-crash) and hopefully it will be included into daily-build binaries from master, and eventually binaries on release branch.

Reply:
Thank you!
Perhaps your patch should fix observed crash (I'm still in progress of building FFmpeg from source code, so will confirm a bit later)

There is another problem which causes that. It is failed "MFXVideoCORE_SetHandle()" call.

I tried to play with device number ("child_device" field), but it doesn't help. Error is raised for non-zero values:
```
[OPENCV:FFMPEG:16] Failed to create Direct3D device
```

Reply:
I discovered one more workaround which allows to run MFX:
- disabling dGPU device through Windows "Device Manager"

This configuration works with enabled OpenCL in OpenCV.

It is really weird that we need to resolve this conflict with NVIDIA dGPU somehow.

---

Current summary for FFMpeg + MFX on Windows:
- &#x2714; works: iGPU only, without OpenCL
- &#x2714; works: iGPU only, with OpenCL
- &#x2714; works: iGPU + dGPU, without OpenCL
- &#x2716; doesn't work: iGPU + dGPU, with OpenCL runtime loaded

Unfortunately the last one configuration is "the default" on such systems.

Reply:
It's strange that OpenCL somehow impacts DirectX/MFX initialization on iGPU + dGPU setup.
I changed how MFX device created, using function `av_hwdevice_ctx_create_derived` instead of `child_device` field. Also in case of D3D11 and VAAPI, the GPU device description now logged via CV_LOG_INFO so indicates GPU vendor/model.
After these changes were done, I've found GPU card around and checked on desktop with iGPU + dGPU, it seems to work ok. In my testing d3d11/mfx initilized on device#1=iGPU, OpenCL initilized on default device#0=dGPU.
```
# example_tapi_video_acceleration.exe -i=INPUT.mp4 -o=output.mp4 -backend=ffmpeg -accel=any -device=1 -opencl=true -out_w=640 -out_h=480
VideoCapture backend = FFMPEG
VideoCapture acceleration = d3d11
VideoCapture acceleration device = 1
OpenCV: FFMPEG: tag 0x34363248/'H264' is not supported with codec id 27 and format 'mp4 / MP4 (MPEG-4 Part 14)'
OpenCV: FFMPEG: fallback to use tag 0x31637661/'avc1'
VideoWriter backend = FFMPEG
VideoWriter acceleration = mfx
VideoWriter acceleration device = 1

Starting frame loop. Press ESC to exit
[ INFO:0] global ..\..\..\modules\core\src\ocl.cpp (1176) cv::ocl::haveOpenCL Initialize OpenCL runtime...
cv::ocl::OpenCLExecutionContext::Impl::getInitializedExecutionContext OpenCL: device=GeForce GTX 980
FPS(last 0.50 sec) = 139.50
FPS(last 0.51 sec) = 134.26
FPS(last 0.51 sec) = 134.48
End of stream
```

Reply:
Thank you for updates!

Validation results on Windows configuration are below. Next steps: check acceleration on Linux.

---

#### Windows+FFmpeg+MFX

FFmpeg crash is gone, but problem with MFX initialization is still here:
```
[ RUN      ] videoio/video_acceleration.write_read/12, where GetParam() = (FOURCC(XVID), .avi, FFMPEG, 30dB, MFX, false)
[AVHWDeviceContext @ 000001ce9dd07400] Error setting child device handle: -6
FFMPEG: VideoWriter acceleration = NONE
[AVHWDeviceContext @ 000001ce9dd07380] Error setting child device handle: -6
FFMPEG: VideoCapture acceleration = NONE
  Real bitrate = 155000 bits-per-second = 0.0201823 bits-per-pixel
  Test finished with min PSNR = 34.5099
[       OK ] videoio/video_acceleration.write_read/12 (680 ms)
```

- without OpenCL it works. Also there are D3D messages about iGPU: `FFMPEG: Using D3D11 video acceleration on GPU device: Intel(R) UHD Graphics`
- with activated OpenCL messages are changed to dGPU: `FFMPEG: Using D3D11 video acceleration on GPU device: NVIDIA GeForce MX230`

---

#### Windows+FFmpeg+D3D

- Local testing works (without OpenCL uses iGPU, with OpenCL uses dGPU).
- There is some issue with D3D11 acceleration on [CI builds](http://pullrequest.opencv.org/buildbot/builders/precommit_custom_windows/builds/1950): `Failed to create Direct3D device`. Perhaps this is related to "Service context" (CI workers launched under Windows service): https://trac.ffmpeg.org/ticket/6827

---

#### Windows+MSMF+D3D11

- MSVS2019: works well
- MSVS2015: D3D11 support is missing during the build:
```
-- Looking for d3d11_4.h
-- Looking for d3d11_4.h - not found
```